builtin/v9/migration/miner.go

```diff
@@ -0,0 +1,141 @@
+	}
+
+	newPrecommitOnChainInfos, err := adt.AsMap(wrappedStore, emptyMap, builtin.DefaultHamtBitwidth)
+	if err != nil {
+		return nil, xerrors.Errorf("failed to load empty map: %w", err)
+	}
+
+	var info miner8.SectorPreCommitOnChainInfo
+	err = oldPrecommitOnChainInfos.ForEach(&info, func(key string) error {
+		var unsealedCid *cid.Cid
+		if len(info.Info.DealIDs) != 0 {
+			pieces := make([]abi.PieceInfo, len(info.Info.DealIDs))
+			for i, dealID := range info.Info.DealIDs {
+				deal, err := m.proposals.GetDealProposal(dealID)
+				if err != nil {
+					return xerrors.Errorf("error getting deal proposal: %w", err)
+				}
+
+				pieces[i] = abi.PieceInfo{
+					PieceCID: deal.PieceCID,
+					Size:     deal.PieceSize,
+				}
+			}
+
+			commd, err := generateUnsealedCID(info.Info.SealProof, pieces)
+			if err != nil {
+				return xerrors.Errorf("failed to generate unsealed CID: %w", err)
+			}
+
+			unsealedCid = &commd
+		}
+
+		err = newPrecommitOnChainInfos.Put(miner9.SectorKey(info.Info.SectorNumber), &miner9.SectorPreCommitOnChainInfo{
+			Info: miner9.SectorPreCommitInfo{
+				SealProof:     info.Info.SealProof,
+				SectorNumber:  info.Info.SectorNumber,
+				SealedCID:     info.Info.SealedCID,
+				SealRandEpoch: info.Info.SealRandEpoch,
+				DealIDs:       info.Info.DealIDs,
+				Expiration:    info.Info.Expiration,
+				UnsealedCid:   unsealedCid,
+			},
+			PreCommitDeposit: info.PreCommitDeposit,
+			PreCommitEpoch:   info.PreCommitEpoch,
+		})
+
+		if err != nil {
+			return xerrors.Errorf("failed to write new precommitinfo: %w", err)
+		}
+
+		return nil
+	})
+
+	if err != nil {
+		return nil, xerrors.Errorf("failed to iterate over precommitinfos: %w", err)
+	}
+
+	newPrecommits, err := newPrecommitOnChainInfos.Root()
+	if err != nil {
+		return nil, xerrors.Errorf("failed to flush new precommits: %w", err)
+	}
+
+	outState := miner9.State{
+		Info:                       inState.Info,
+		PreCommitDeposits:          inState.PreCommitDeposits,
+		LockedFunds:                inState.LockedFunds,
+		VestingFunds:               inState.VestingFunds,
+		FeeDebt:                    inState.FeeDebt,
+		InitialPledge:              inState.InitialPledge,
+		PreCommittedSectors:        newPrecommits,
+		PreCommittedSectorsCleanUp: inState.PreCommittedSectorsCleanUp,
+		AllocatedSectors:           inState.AllocatedSectors,
+		Sectors:                    inState.Sectors,
+		ProvingPeriodStart:         inState.ProvingPeriodStart,
+		CurrentDeadline:            inState.CurrentDeadline,
+		Deadlines:                  inState.Deadlines,
+		EarlyTerminations:          inState.EarlyTerminations,
+		DeadlineCronActive:         inState.DeadlineCronActive,
+	}
+
+	newHead, err := store.Put(ctx, &outState)
+	return &actorMigrationResult{
+		newCodeCID: m.migratedCodeCID(),
+		newHead:    newHead,
+	}, err
+}
+
+func generateUnsealedCID(proofType abi.RegisteredSealProof, pieces []abi.PieceInfo) (cid.Cid, error) {
```

Comment:
This method became a lot smaller after I took out the padding logic (which the commp util already does). I'll flatten it into the caller.

Reply:
They are like 1:1 identical signatures now no?

Reply:
Yes they are.